### PR TITLE
Fast iterative Feature adversaries

### PR DESCRIPTION
**Pull request by fartashf**

### CODE REVIEW

File: examples/imagenet_featadvs/attack_model_featadv.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
+"""Runs CleverHans attacks on the Madry Lab MNIST challenge model
```
Comment (goodfeli) [inline examples/imagenet_featadvs/attack_model_featadv.py]: If this is for the MNIST challenge, why is it in a directory with "imagenet" in the name?

File: examples/imagenet_featadvs/attack_model_featadv.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import time
+import numpy as np
+
+import tensorflow as tf
+# from tensorflow.examples.tutorials.mnist import input_data
```
Comment (goodfeli) [inline examples/imagenet_featadvs/attack_model_featadv.py]: Let's not commit commented code

File: examples/imagenet_featadvs/attack_model_featadv.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                                                  test_end=test_end)
+
+    assert Y_train.shape[1] == 10.
+
+    # NOTE: for compatibility with Madry Lab downloadable checkpoints,
+    # we cannot enclose this in a scope or do anything else that would
+    # change the automatic naming of the variables.
+    model = MadryMNIST()
+
+    # x_input = tf.placeholder(tf.float32, shape=[None, 784])
```
Comment (goodfeli) [inline examples/imagenet_featadvs/attack_model_featadv.py]: Let's not commit commented code

File: examples/imagenet_featadvs/attack_model_featadv.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        ffg = FastIterativeFeatureAdversaries(model)
+        ffg_params = {'eps': 0.3, 'clip_min': 0., 'clip_max': 1.,
+                      'nb_iter': FLAGS.nb_iter, 'eps_iter': 0.1,
+                      'layer': FLAGS.layer,
+                      'targeted': True}  # , 'ord': 2}
+        g = tf.abs(tf.truncated_normal(tf.shape(x_image), 1./2))
+        adv_x = ffg.generate(x_image, g, **ffg_params)
+    elif FLAGS.attack_type == 'ffg_untarg':
+        print("FFG")
+        # ['h_conv1', 'h_conv2', 'h_fc1', 'h_pool1', 'h_pool2', 'h_pool2_flat',
```
Comment (goodfeli) [inline examples/imagenet_featadvs/attack_model_featadv.py]: More dead code

File: examples/imagenet_featadvs/attack_model_featadv.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        'label_smooth', 0.1,
+        "Amount to subtract from correct label"
+        " and distribute among other labels")
+    flags.DEFINE_string(
+        'attack_type', 'fgsm',
+        "Attack type: 'fgsm'->fast gradient sign method,"
+        " 'bim'->'basic iterative method'")
+    flags.DEFINE_string('checkpoint_dir', '',  # default_checkpoint_dir,
+                        'Checkpoint directory to load')
+    # ['h_conv1', 'h_conv2', 'h_fc1', 'h_pool1', 'h_pool2', 'h_pool2_flat',
```
Comment (goodfeli) [inline examples/imagenet_featadvs/attack_model_featadv.py]: Dead code

**Comment by goodfeli** [comment]
It looks like there aren't any tests, can you add at least one test per new function / class?

File: cleverhans/feat_advs.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+import tensorflow as tf
+
+from cleverhans.attacks import Attack
+from cleverhans.model import Model
+
+
+def ffg(x, s_feat, g_feat, eps=0.3, ord=np.inf,
+        clip_min=None, clip_max=None, targeted=False):
+    """
+    TensorFlow implementation of the Fast Gradient Method.
```
Comment (goodfeli) [inline cleverhans/feat_advs.py]: This docstring seems wrong; presumably this isn't the Fast Gradient Method

File: cleverhans/feat_advs.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
+import numpy as np
```
Comment (goodfeli) [inline cleverhans/feat_advs.py]: Module level docstring

Comment (goodfeli) [reply cleverhans/feat_advs.py]: You should put a link to the paper in here somewhere, or I think most people won't know what this is.

Also, for your sake, you probably want to give people instructions for citing the paper :)

File: cleverhans/feat_advs.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+class FastFeatureGradient(Attack):
+
+    """
+    This is similar to FGSM but done for the internal representations.
+    """
+
+    def __init__(self, model, back='tf', sess=None):
+        """
+        Create a FastGradientMethod instance.
```
Comment (yenchenlin) [inline cleverhans/feat_advs.py]: This should be `Create a FastFeatureGradient instance`, right?

**Comment by fartashf** [comment]
Yeah. Thanks. This PR is unfinished. I'm resolving the issues.

File: cleverhans/feat_advs.py (hunk 7 above)
Comment (fartashf) [reply cleverhans/feat_advs.py]: Docstring added in 4ec1f53. Regarding citation, is it enough to include the link to arxiv?

File: cleverhans/feat_advs.py (hunk 6 above)
Comment (fartashf) [reply cleverhans/feat_advs.py]: Fixed in 7f5b62d and 9459f33.

File: cleverhans/feat_advs.py (hunk 8 above)
Comment (fartashf) [reply cleverhans/feat_advs.py]: Fixed in c80b238. Thank you.

File: cleverhans/feat_advs.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+                    Possible values: np.inf, 1 or 2.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        import tensorflow as tf
+
+        # Parse and save attack-specific parameters
+        assert self.parse_params(**kwargs)
+
+        def clip_eta(eta):
```
Comment (yenchenlin) [inline cleverhans/feat_advs.py]: I think you can use `clip_eta` function defined in `utils_tf.py` [here](https://github.com/tensorflow/cleverhans/blob/5114ebd6b54c4197819042a7d716b2ac4e8285a0/cleverhans/utils_tf.py#L396-L421).

File: examples/imagenet_featadvs/attack_model_featadv.py (hunk 1 above)
Comment (fartashf) [reply examples/imagenet_featadvs/attack_model_featadv.py]: Fixed in 47b04aa.

File: examples/imagenet_featadvs/attack_model_featadv.py (hunk 2 above)
Comment (fartashf) [reply examples/imagenet_featadvs/attack_model_featadv.py]: Fixed in 47b04aa.

File: examples/imagenet_featadvs/attack_model_featadv.py (hunk 4 above)
Comment (fartashf) [reply examples/imagenet_featadvs/attack_model_featadv.py]: Fixed in 47b04aa.

File: examples/imagenet_featadvs/attack_model_featadv.py (hunk 5 above)
Comment (fartashf) [reply examples/imagenet_featadvs/attack_model_featadv.py]: Fixed in 47b04aa.

File: cleverhans/feat_advs.py (hunk 9 above)
Comment (fartashf) [reply cleverhans/feat_advs.py]: Fixed in b73717bdc1206508d1e425fd84a5b1298ba59705.

**Comment by fartashf** [comment]
Thanks for the comments. I have addressed the comments except for the one regarding writing tests. The example in `imagenet_featadvs/` is supposed to be a very basic test which is not converging to 0 right now. Let me know if you have any ideas for unit tests and if you found any bugs through reviewing.

**Comment by goodfeli** [comment]
The tests are failing with "from utils_tf import clip_eta" getting an ImportError

File: cleverhans/feat_advs.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+        # feat.shape = (batch, c) or (batch, w, h, c)
+        axis = range(1, len(a_feat.shape))
+
+        # Compute loss
+        # This is a targeted attack, hence the negative sign
+        loss = -tf.reduce_sum(tf.square(a_feat - g_feat), axis)
+
+        # Define gradient of loss wrt input
+        grad, = tf.gradients(loss, x)
```
Comment (npapernot) [inline cleverhans/feat_advs.py]: Shouldn't this be `adv_x` here instead of `x`?

File: cleverhans/feat_advs.py (hunk 7 above)
Comment (npapernot) [reply cleverhans/feat_advs.py]: Yes it should be fine to add the link, at least this is what we have done so far with other attacks :)

File: cleverhans/feat_advs.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        self.eps_iter = eps_iter
+        self.nb_iter = nb_iter
+        self.ord = ord
+        self.clip_min = clip_min
+        self.clip_max = clip_max
+
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, 1, 2]:
+            raise ValueError("Norm order must be either np.inf, 1, or 2.")
+        if self.back == 'th':
```
Comment (npapernot) [inline cleverhans/feat_advs.py]: You can remove this given that we dropped Theano support.

**Comment by npapernot** [comment]
@fartashf are you able to take a look at this PR in the next couple of days?

File: cleverhans/feat_advs.py (hunk 11 above)
Comment (fartashf) [reply cleverhans/feat_advs.py]: Removed in 30b4472.

File: cleverhans/feat_advs.py (hunk 10 above)
Comment (fartashf) [reply cleverhans/feat_advs.py]: Thanks for catching that. Fixed in 11919a2.

**Comment by fartashf** [comment]
@npapernot thanks for the reminder. I fixed the issues and added a test. The tests fail for test_mnist_tutorial_tf.TestMNISTTutorialTF. I guess it's again due to a tight threshold.

**Comment by goodfeli** [comment]
Quoting the failure here:
    Traceback (most recent call last):
      File "/home/travis/build/tensorflow/cleverhans/tests_tf/test_mnist_tutorial_tf.py", line 27, in test_mnist_tutorial_tf
        self.assertLess(report.train_clean_train_adv_eval, 0.036)
    AssertionError: 0.036799999999999999 not less than 0.036

File: tests_tf/test_attacks.py [hunk 12]
```diff
@@ -777,5 +778,82 @@ def test_multiple_initial_random_step(self):
         self.assertTrue(np.mean(orig_labs == new_labs_multi) < 0.1)
 
 
+class TestFastFeatureAdversaries(CleverHansTest):
+    def setUp(self):
+        super(TestFastFeatureAdversaries, self).setUp()
+        import tensorflow as tf
+
+        def make_imagenet_cnn(input_shape=(None, 224, 224, 3)):
```
Comment (goodfeli) [inline tests_tf/test_attacks.py]: How long does this test take to run? The model looks pretty big for a Travis test

**Comment by goodfeli** [comment]
Since this changes `FastGradientMethod` it does seem possible the failure is real.

Notes for my future self / other people following the thread: this PR changes `FastGradientMethod` to call `clip_eta` rather than having its own version of the `clip_eta` code internally. This is a great refactor (thanks Fartash!) but I'm concerned that there's something subtly different between the two versions that we don't know about.

Fartash: are the tests failing on your machine too or just on Travis? It looks like they fail only for newer version of TensorFlow. Do they fail consistently or flakily? From one run of Travis it looks like we have 4 trials that all support the hypothesis that the tests fail consistently for TensorFlow > 1.1. Since they consistently passed before, this looks like something did meaningfully change in `clip_eta`.

**Comment by fartashf** [comment]
My bad for mixing multiple PRs. This PR is actually not changing the `FastGradientMethod`. It only refactors `BasicIterativeMethod` to use `clip_eta`. This should not affect the tutorial. `clip_eta` was introduced in 47d85a6d73401d60f21be0653bf7bae78b4c459b and later was moved to `utils_tf.py` in 766add490bef3fc48393b4056973f62ace3e1f55.

Tests don't fail on my machine. I tested with Tensorflow 1.4.0 using python 2.7.14 and 3.6.3.

File: tests_tf/test_attacks.py (hunk 12 above)
Comment (goodfeli) [reply tests_tf/test_attacks.py]: 19 seconds isn't so bad. Is there anything trivial that can speed it up? For example, it looks like the batch size is 100. Does the test still work with batch size 1?

Does the attack still succeed on unrealistically small models, e.g. with only 32 filters per layer?

**Comment by goodfeli** [comment]
Are you sure the test only takes 19 seconds? The tests for this PR took up to 33 minutes to run but if I look at another currently open PR they're consistently under 26.

https://travis-ci.org/tensorflow/cleverhans/builds/314515713?utm_source=github_status&utm_medium=notification

**Comment by goodfeli** [comment]
This one is also just barely over 26 for only one of 4 runs: https://travis-ci.org/tensorflow/cleverhans/builds/314921361?utm_source=github_status&utm_medium=notification

File: tests_tf/test_attacks.py (hunk 12 above)
Comment (fartashf) [reply tests_tf/test_attacks.py]: Changed 100 to 10 in 58a7054 and it's faster now. Thanks for the suggestion.

**Comment by goodfeli** [comment]
Thanks, it does look a lot faster now.

Unfortunately, we now have two more examples of the tensorflow 1.1 tests failing, and by exactly the same amount of error as before:

Traceback (most recent call last):
  File "/home/travis/build/tensorflow/cleverhans/tests_tf/test_mnist_tutorial_tf.py", line 27, in test_mnist_tutorial_tf
    self.assertLess(report.train_clean_train_adv_eval, 0.036)
AssertionError: 0.036799999999999999 not less than 0.036

Since this isn't happening to pull requests in other branches, I do think there's something unusual about this branch.

One thing you might try is to rebase your branch on top of master and run the tests locally. If the rebase gets the tests to fail locally, that would give you some more convenient signal for debugging. You'll need to push -f back to the pull request after the merge.

**Comment by fartashf** [comment]
I rebased and the tests still pass on my machine.
To make the PR simpler, I separated the part regarding `clip_eta` into a new PR #326. We will get a clue after that one is tested.

**Comment by goodfeli** [comment]
Thanks! Feel free to push -f the remainder of the code to this branch. If we're lucky, it will just pass.

**Comment by fartashf** [comment]
The checks have passed. I think it is only because of randomness in the tests. The only change I made was moving part of the code to PR #326.

**Comment by goodfeli** [comment]
Thanks!

**Pull request merged**